src/levanter/models/gemma.py

```diff
@@ -0,0 +1,608 @@
+        k_t, k_emb = jrandom.split(key, 2)
+        transformer = GemmaTransformer.init(config, key=k_t)
+        embeddings = GemmaEmbedding.init(Vocab, config, key=k_emb)
+        lm_head = hnn.Linear.init(In=config.Embed, Out=Vocab, key=k_emb, use_bias=False, out_first=True)
+        return GemmaLMHeadModel(transformer, embeddings, lm_head)
+
+    def __call__(
+        self,
+        input_ids: NamedArray,
+        attn_mask: Optional[Union[NamedArray, AttentionMask]] = None,
+        *,
+        key=None,
+    ) -> NamedArray:
+        """
+        Args:
+            input_ids (NamedArray): [batch, position]
+                Indices of input sequence tokens in the vocabulary.
+            attn_mask (Union[NamedArray, AttentionMask], optional): [batch, position]
+                Mask to avoid performing attention on the padding token indices of the encoder input.
+                The attn_mask from training pipeline may be an AttentionMask object instead of NamedArray
+        """
+        k_t, k_head = maybe_rng_split(key, 2)
+        x = self.embeddings.embed(input_ids)
+        x = self.transformer(x, attn_mask=attn_mask, key=k_t)
+        lm_logits = self.lm_head(x, key=k_head)
+        return lm_logits
+
+    def resize_vocab(self, new_size: int, key=None) -> "LmHeadModel[GemmaConfig]":
+        new_Vocab = self.Vocab.resize(new_size)
+        k1, k2 = maybe_rng_split(key, 2)
+        new_embeddings = self.embeddings.resize_embeddings(new_size, key=k1)
+        new_lm_matrix = hax.tree_util.resize_axis(self.lm_head.weight, self.Vocab, new_size, key=k2)
+        new_lm_head = dataclasses.replace(self.lm_head, Out=new_Vocab, weight=new_lm_matrix)
+
+        return dataclasses.replace(self, embeddings=new_embeddings, lm_head=new_lm_head)
+
+    def _state_dict_key_map(self) -> Dict[str, Optional[str]]:
+        """Map from Levanter model names to HF."""
+        return {"transformer": "model", "embeddings": None}
+
+    def from_state_dict(self, state_dict: StateDict, prefix: Optional[str] = None):
+        # unflatten the linear layers of HF state_dict to match the shape of GemmaMlp
+        d = state_dict.copy()
+
+        # Gemma lm_head weights are tied (shared) with the embedding layer.
+        d["lm_head.weight"] = d["model.embed_tokens.weight"]
+
+        d.update(
+            unflatten_linear_layers(apply_prefix(prefix, "lm_head"), d, self.lm_head, out_dims_first_in_dict=True)
+        )
+        return super().from_state_dict(d, prefix)
+
+    def update_state_dict(self, state_dict: StateDict, prefix: Optional[str] = None) -> StateDict:
+        my_dict: StateDict = {}
+        super().update_state_dict(my_dict, prefix=prefix)
+
+        my_dict.update(
+            flatten_linear_layers(apply_prefix(prefix, "lm_head"), self.lm_head, out_dims_first_in_dict=True)
+        )
+
+        state_dict.update(my_dict)
+        return state_dict
+
+
+def _rotate_half(x: NamedArray) -> NamedArray:
```

Comment:
what are your thoughts on this kind of copy-paste? HF has an explicit "no code reuse" policy, and there are extremes in the other direction (MaxText is very shared-code happy, and xtransformers even more so). 

My inclination is for things that are literally the same like this to be shared, but if there's even a little bit of mismatch, to not share. (I.e. no adding flags just to make things shared). So I'm kinda inclined to say we should move these functions from llama to their own file and call them in both places. WDYT?

Reply:
yeah I like that. I'm not as certain about the DecoderLayer and in particular I worry that sometimes it will make HF compatibility harder (breaking the ~isomorphism to the HF structure will complicate export serialization)